stdlib/src/ssl/cert.rs

```diff
@@ -1,232 +1,1427 @@
+            invalid_data(format!("no start line: {}", context))
+        }
+
+        pub fn parse_failed(e: impl std::fmt::Display) -> io::Error {
+            invalid_data(format!("Failed to parse PEM certificate: {}", e))
+        }
+
+        pub fn parse_failed_debug(e: impl std::fmt::Debug) -> io::Error {
+            invalid_data(format!("Failed to parse PEM certificate: {:?}", e))
+        }
+
+        pub fn invalid_cert() -> io::Error {
+            invalid_data("No certificates found in certificate file")
+        }
+    }
+
+    /// DER parsing error variants
+    pub mod der {
+        use super::*;
+
+        pub fn not_enough_data(context: &str) -> io::Error {
+            invalid_data(format!("not enough data: {}", context))
+        }
+
+        pub fn parse_failed(e: impl std::fmt::Display) -> io::Error {
+            invalid_data(format!("Failed to parse DER certificate: {}", e))
+        }
+    }
+
+    /// Private key error variants
+    pub mod key {
+        use super::*;
+
+        pub fn not_found(context: &str) -> io::Error {
+            invalid_data(format!("No private key found in {}", context))
+        }
+
+        pub fn parse_failed(e: impl std::fmt::Display) -> io::Error {
+            invalid_data(format!("Failed to parse private key: {}", e))
+        }
+
+        pub fn parse_encrypted_failed(e: impl std::fmt::Display) -> io::Error {
+            invalid_data(format!("Failed to parse encrypted private key: {}", e))
+        }
+
+        pub fn decrypt_failed(e: impl std::fmt::Display) -> io::Error {
+            io::Error::other(format!(
+                "Failed to decrypt private key (wrong password?): {}",
+                e
+            ))
+        }
+    }
+
+    /// Convert error message to rustls::Error with InvalidCertificate wrapper
+    pub fn to_rustls_invalid_cert(msg: impl Into<String>) -> rustls::Error {
+        rustls::Error::InvalidCertificate(rustls::CertificateError::Other(rustls::OtherError(
+            Arc::new(invalid_data(msg)),
+        )))
+    }
+}
+
+// Helper Functions for Certificate Parsing
+
+/// Map X.509 OID to human-readable attribute name
+///
+/// Converts common X.509 Distinguished Name OIDs to their standard names.
+/// Returns the OID string itself if not recognized.
+fn oid_to_attribute_name(oid_str: &str) -> &str {
+    match oid_str {
+        "2.5.4.3" => "commonName",
+        "2.5.4.6" => "countryName",
+        "2.5.4.7" => "localityName",
+        "2.5.4.8" => "stateOrProvinceName",
+        "2.5.4.10" => "organizationName",
+        "2.5.4.11" => "organizationalUnitName",
+        "1.2.840.113549.1.9.1" => "emailAddress",
+        _ => oid_str,
+    }
+}
+
+/// Format IP address (IPv4 or IPv6) to string
+///
+/// Formats raw IP address bytes according to standard notation:
+/// - IPv4: dotted decimal (e.g., "192.0.2.1")
+/// - IPv6: colon-separated hex (e.g., "2001:DB8:0:0:0:0:0:1")
+fn format_ip_address(ip: &[u8]) -> String {
```

Comment:
Wouldn't make more sense to use one of:
- https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html
- https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html